FogCreek/Glitch-Community
# Profuse tea: Rework webpack and ssr transpile for faster startup

## Links
https://profuse-tea.glitch.me

## Changes:
- When `DEPLOY_ENV` is set to `production` the server expects the build to be managed elsewhere
  - Webpack is run via `prestart` in watch mode (rather than being spawned by the server code, so editing the server doesn't cause the client to rebuild from scratch!)
  - A babel transpile is started in watch mode for ssr code (rather than using babel register)
  - This ensures that the server code waits for nothing and can immediately server requests using the last build
- When `DEPLOY_ENV` is set to `ci` the server acts the same as above but the two watchers do not get started, so it can serve a static build instead
- When `DEPLOY_ENV` is set to anything else, the site stalls requests in order to use the latest code
  - Webpack is run in the server process using webpack-dev-middleware
  - Babel is run in the server process using babel register (like it is on prod now)
- The client code is built into build/client instead of build/, so that webpack doesn't clean up files from other builds
- I moved shared/aliases.js to the root folder because it's used by webpack and eslint, so the server needs to restart when it is changed
- I told webpack not to output the name/path of every file it parses because it floods the logs

## How To Test:
-Try running the stop command against the project and see how quickly you can reconnect. (I'm averaging about 30s but sometimes it hits me with a pnpm reinstall and takes a minute or so)
- Join the project and make edits in different places and make sure the watchers all still do what they should

## Feedback I'm looking for:
- `DEPLOY_ENV` controls how the site manages builds, should it get renamed to something like `BUILD_TYPE=static|watcher|dev`?
- I don't have any handling for running on prod before the initial ssr build finishes. it'll just fail to render. What should it do, if anything else?
